benchmark/llama.py

```diff
@@ -0,0 +1,187 @@
+            (benchmark_id, cpu_util, mem_megabytes, gpu_util, gpu_mem_megabytes),
+        )
+        sleep(0.01)
+        conn.commit()
+    conn.close()
+
+
+def run_benchmark(branch: str, commit_id: str, commit_msg: str):
+    continue_metric_collection = Event()
+    gpu_stats = gpustat.GPUStatCollection.new_query()
+    gpu_name = gpu_stats[0]["name"]
+    conn = psycopg2.connect("dbname=metrics host=localhost user=postgres password=postgres")
+    cur = conn.cursor()
+    cur.execute(
+        "INSERT INTO benchmarks (branch, commit_id, commit_msg, gpu_name) VALUES (%s, %s, %s, %s) RETURNING benchmark_id",
+        (branch, commit_id, commit_msg, gpu_name),
+    )
+    conn.commit()
+    benchmark_id = cur.fetchone()[0]
+    metrics_thread = Thread(target=collect_metrics, args=[benchmark_id, continue_metric_collection])
+    metrics_thread.start()
+
+    try:
+        os.environ["TOKENIZERS_PARALLELISM"] = "false"  # silence warnings when compiling
+
+        device = "cuda"
+        ckpt = "bert-base-uncased"
+
+        # This is to avoid counting download in model load time measurement
+        model = AutoModelForCausalLM.from_pretrained(ckpt, torch_dtype=torch.float16)
+        start = perf_counter()
+        model = AutoModelForCausalLM.from_pretrained(ckpt, torch_dtype=torch.float16)
+        model.to(device)
+        end = perf_counter()
+        model_load_time = end - start
+
+        tokenizer = AutoTokenizer.from_pretrained(ckpt)
+
+        prompt = "Why dogs are so cute?"
+        inputs = tokenizer(prompt, return_tensors="pt").to(device)
+
+        # Specify the max length (including both the prompt and the response)
+        # When calling `generate` with `cache_implementation="static" later, this is also used to create a `StaticCache` object
+        # with sequence length = `max_length`. The longer the more you will re-use it
+        model.generation_config.max_length = 128
+
+        # without `torch.compile`: each call takes ~ 5.0 seconds (on A100 80G + torch 2.3)
+        start = perf_counter()
+        model.generate(**inputs, do_sample=False)
+        end = perf_counter()
+        first_eager_fwd_pass_time = end - start
+        # TODO:
+        # response = tokenizer.batch_decode(outputs)[0]
+        start = perf_counter()
+        model.generate(**inputs, do_sample=False)
+        end = perf_counter()
+        second_eager_fwd_pass_time = end - start
+
+        # `torch.compile(model, ...)` is not recommended as you compile callbacks
+        # and full generate. We recommend compiling only the forward for now.
+        # "reduce-overhead" will use cudagraphs.
+        model.forward = torch.compile(model.forward, mode="reduce-overhead", fullgraph=True)
+        model.generation_config.cache_implementation = "static"
+
+        # with `torch.compile` (on A100 80G + torch 2.3)
+        # 1st call: ~ 90 seconds
+        start = perf_counter()
+        model.generate(**inputs, do_sample=False)
```

Comment:
okay, we are more computing the full generation here ! let's keep this . You can copy past the entire `run_benchmark` and use this: https://gist.github.com/ArthurZucker/5dc54a3fb443e979fac437e5df7c800b 
which is the proper forward calling!